### PR TITLE
kubectl-karmada add `init` Command

### LINKED ISSUE #804: Add `karmadm` tool init karmada and join/unjoin member cluster
**What would you like to be added**:
`karmadm` is specifically used to initialize Karmada and join/unjoin member cluster or other preparatory work.

**Why is this needed**:
`karmadactl` is used to operate object in karmada control plane. Because Karmada initialization requires working with multiple clusters, which can be isolated to `karmadm`.

### PR DESCRIPTION
**Pull request by prodanlabs**
**What type of PR is this?**
/kind design

**What this PR does / why we need it**:
easy and quick install of karmada
**Which issue(s) this PR fixes**:
Fixes #804

**Special notes for your reviewer**:
kubectl-karmada add `init` Command
```
# ./kubectl-karmada
kubectl karmada controls a Kubernetes Cluster Federation.

Usage:
  karmada [flags]
  karmada [command]

Available Commands:
  completion  generate the autocompletion script for the specified shell
  cordon      Mark cluster as unschedulable
  get         Display one or many resources
  help        Help about any command
  init        bootstrap install karmada (default in kubernetes)
  join        Register a cluster to control plane
  taint       Update the taints on one or more clusters.
  uncordon    Mark cluster as schedulable
  unjoin      Remove the registration of a cluster from control plane
  version     Print the version information.
```
**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by prodanlabs** [comment]
Hi @RainbowMango
I added a sub-command to kubectl-karmada. If there is no problem, the new sub-commands will also be added in this form.

**Comment by lonelyCZ** [comment]
It looks good.  I was interested in this feature in #804. Thank you for implementing it.

I have considered whether we can read the crd and other object  from local file while compiling it, rather than setting it a constant like `pkg/kaadm/karmada/crds/bases_cluster_karmada_io_clusters.go`, because we possibly update these object, which could increase code maintenance costs. What do you think?

**Comment by prodanlabs** [comment]
@lonelyCZ  Good idea. I'm still working on the final work of `karmada in container`, let me see if I have time later

**Comment by prodanlabs** [comment]
Hi @RainbowMango trouble re-trigger `ci`

**Comment by RainbowMango** [comment]
Done. Sorry for the delay.

**Comment by prodanlabs** [comment]
> 完毕。抱歉耽搁了。

Unfortunately, it needs to be repeated again.
Thanks community, let me regain the golang specification :)

**Comment by prodanlabs** [comment]
> Done. Sorry for the delay.

Unfortunately, it needs to be repeated again.
Thanks community, let me regain the golang specification :)

**Comment by lonelyCZ** [comment]
Hello, @prodanlabs , perhaps you can activate ci test in your github fork (github action).

Also, please squash your commits to one, and sign your commit according to [https://github.com/karmada-io/karmada/pull/1094/checks?check_run_id=4498109394](https://github.com/karmada-io/karmada/pull/1094/checks?check_run_id=4498109394)

**Comment by prodanlabs** [comment]
thank @lonelyCZ  @RainbowMango , I have benefited a lot today.

Submit PR after passing the next inspection.

**Comment by RainbowMango** [comment]
> It looks good. I was interested in this feature in #804. Thank you for implementing it.

Yeah. I added `Fixes #804` to this PR description.

File: pkg/kaadm/kaadm.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+// NewCmdInit init karmada.
+func NewCmdInit() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "init",
+		Short: "bootstrap install karmada (default in kubernetes)",
+		Long:  `Installation options.`,
+		Run: func(cmd *cobra.Command, args []string) {
+			kubernetes.Deploy()
+		},
+		Example: "kubectl karmada init --master=xxx.xxx.xxx.xxx",
```
Comment (RainbowMango) [inline pkg/kaadm/kaadm.go:19]: If we can give more examples would be wonderful!
Like [karmada get](https://github.com/karmada-io/karmada/blob/006bccd0bdc992963c1e850fb3d5668311964088/pkg/karmadactl/get.go#L543-L558) or [kubeadm](https://github.com/kubernetes/kubernetes/blob/c83a94da72416219cf9f022dc44b0e6db158f092/cmd/kubeadm/app/cmd/cmd.go#L35-L66).

**Comment by RainbowMango** [comment]
I run it locally and looks great, sharing the steps here:
```
// create a cluster by 'kind'
# kind create cluster

// show the cluster
# docker ps 
CONTAINER ID        IMAGE                  COMMAND                  CREATED             STATUS              PORTS                       NAMES
96d3cf95ecfa        kindest/node:v1.21.1   "/usr/local/bin/entr…"   4 minutes ago       Up 4 minutes        127.0.0.1:42174->6443/tcp   kind-control-plane
1334336ee8e6        registry:2             "/entrypoint.sh /etc…"   6 months ago        Up 2 weeks          127.0.0.1:5000->5000/tcp    kind-registry

// build kubectl-karmada
# make kubectl-karmada

// copy kubectl-karmada to cluster node(skip this if not kind cluster)
# docker cp kubectl-karmada kind-control-plane:/

// enter the kind container
# docker exec -it kind-control-plane bash

// install karmada
# /kubectl-karmada init --master=172.18.0.2 --crds https://github.com/karmada-io/karmada/releases/download/v0.10.1/crds.tar.gz --kubeconfig /etc/kubernetes/admin.conf 
I1216 07:37:45.862959    4256 cert.go:230] Generate ca certificate success.
I1216 07:37:46.000798    4256 cert.go:230] Generate etcd-server certificate success.
I1216 07:37:46.413977    4256 cert.go:230] Generate etcd-client certificate success.
I1216 07:37:46.717352    4256 cert.go:230] Generate karmada certificate success.
I1216 07:37:46.717377    4256 deploy.go:103] crds file name: /var/lib/karmada/crds.tar.gz
Downloading...[ 100.00% ]
Downloading...[ 100.00% ]
Download complete.I1216 07:37:47.635042    4256 deploy.go:170] kubeconfig file: /etc/kubernetes/admin.conf
I1216 07:37:47.638361    4256 deploy.go:188] master ip: [172.18.0.2]
I1216 07:37:47.639509    4256 deploy.go:347] Create karmada kubeconfig success.
I1216 07:37:47.654572    4256 namespace.go:38] Create Namespace 'karmada-system' successfully.
I1216 07:37:48.043741    4256 secret.go:78] secret kubeconfig Create successfully.
I1216 07:37:48.450723    4256 secret.go:78] secret etcd-cert Create successfully.
I1216 07:37:48.843725    4256 secret.go:78] secret karmada-cert Create successfully.
I1216 07:37:49.242543    4256 secret.go:78] secret karmada-webhook-cert Create successfully.
I1216 07:37:50.244969    4256 services.go:66] service etcd create successfully.
I1216 07:37:50.244989    4256 deploy.go:261] create etcd StatefulSets
I1216 07:37:50.646409    4256 check.go:100] etcd desired replicaset is 1, currently: 1
W1216 07:37:53.656715    4256 check.go:54] pod: etcd-0 not ready. status: PodInitializing
W1216 07:37:54.660289    4256 check.go:54] pod: etcd-0 not ready. status: PodInitializing
W1216 07:37:55.660175    4256 check.go:54] pod: etcd-0 not ready. status: PodInitializing
W1216 07:37:56.659252    4256 check.go:54] pod: etcd-0 not ready. status: PodInitializing
W1216 07:37:57.659835    4256 check.go:54] pod: etcd-0 not ready. status: PodInitializing
W1216 07:37:58.662838    4256 check.go:54] pod: etcd-0 not ready. status: PodInitializing
W1216 07:37:59.660575    4256 check.go:54] pod: etcd-0 not ready. status: PodInitializing
W1216 07:38:01.903588    4256 check.go:54] pod: etcd-0 not ready. status: PodInitializing
I1216 07:38:02.659640    4256 check.go:51] pod: etcd-0 is ready. status: Running
I1216 07:38:02.659673    4256 deploy.go:244] create karmada ApiServer Deployment
I1216 07:38:02.670954    4256 services.go:66] service karmada-apiserver create successfully.
W1216 07:38:05.887888    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: ContainerCreating
W1216 07:38:06.890489    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:07.891001    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:08.890643    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:09.891045    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:10.891697    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:11.890877    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:12.890698    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:13.890927    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:14.890827    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:15.890826    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:16.891054    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:17.891046    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:18.890344    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:19.890885    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:20.891164    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:21.891007    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:22.890752    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:23.891167    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:24.890663    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:25.890867    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:26.890930    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:27.890795    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:28.891007    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:29.890849    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:30.891052    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
W1216 07:38:31.891422    4256 check.go:54] pod: karmada-apiserver-5899b8fbbf-8n2ms not ready. status: Running
I1216 07:38:32.890372    4256 check.go:51] pod: karmada-apiserver-5899b8fbbf-8n2ms is ready. status: Running
I1216 07:38:32.904289    4256 deploy.go:55] Initialize karmada bases crd resource `/var/lib/karmada/crds/bases`
I1216 07:38:33.311235    4256 deploy.go:66] Initialize karmada patches crd resource `/var/lib/karmada/crds/patches`
I1216 07:38:33.722670    4256 deploy.go:79] Crate MutatingWebhookConfiguration mutating-config.
I1216 07:38:33.735019    4256 deploy.go:84] Crate ValidatingWebhookConfiguration validating-config.
I1216 07:38:33.740571    4256 deploy.go:276] create karmada kube controller manager Deployment
I1216 07:38:33.749452    4256 services.go:66] service kube-controller-manager create successfully.
W1216 07:38:36.760013    4256 check.go:54] pod: kube-controller-manager-7bb96b7779-8tc94 not ready. status: ContainerCreating
I1216 07:38:37.763666    4256 check.go:51] pod: kube-controller-manager-7bb96b7779-8tc94 is ready. status: Running
I1216 07:38:37.763688    4256 deploy.go:289] create karmada scheduler Deployment
I1216 07:38:40.774411    4256 check.go:51] pod: karmada-scheduler-7b9d8b5764-mz55q is ready. status: Running
I1216 07:38:40.774432    4256 deploy.go:299] create karmada controller manager Deployment
I1216 07:38:43.782370    4256 check.go:51] pod: karmada-controller-manager-556cf896bc-75htm is ready. status: Running
I1216 07:38:43.782400    4256 deploy.go:309] create karmada webhook Deployment
I1216 07:38:43.790436    4256 services.go:66] service karmada-webhook create successfully.
I1216 07:38:46.799511    4256 check.go:51] pod: karmada-webhook-7cf7986866-cjxmz is ready. status: Running
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
| Push mode                                                                                                                                                                    |
|                                                                                                                                                                              |
| Step 1: Member kubernetes join karmada control plane                                                                                                                         |
|                                                                                                                                                                              |
| (In karmada)~#  cat ~/.kube/config  | grep current-context | sed 's/: /\n/g'| sed '1d' #MEMBER_CLUSTER_NAME                                                                  |
| (In karmada)~# kubectl-karmada  --kubeconfig /var/lib/karmada/karmada-apiserver.config  join ${MEMBER_CLUSTER_NAME} --cluster-kubeconfig=$HOME/.kube/config                  |
|                                                                                                                                                                              |
| Step 2: Create member kubernetes kubeconfig secret                                                                                                                           |
|                                                                                                                                                                              |
| (In member kubernetes)~# kubectl create ns karmada-system                                                                                                                    |
| (In member kubernetes)~# kubectl create secret generic ${MEMBER_CLUSTER_NAME}-kubeconfig --from-file=${MEMBER_CLUSTER_NAME}-kubeconfig=$HOME/.kube/config  -n karmada-system |
|                                                                                                                                                                              |
| Step 3: Create karmada scheduler estimator                                                                                                                                   |
|                                                                                                                                                                              |
| (In member kubernetes)~# sed -i "s/{{member_cluster_name}}/${MEMBER_CLUSTER_NAME}/g" /var/lib/karmada/karmada-scheduler-estimator.yaml                                       |
| (In member kubernetes)~# kubectl create -f  /var/lib/karmada/karmada-scheduler-estimator.yaml                                                                                |
|                                                                                                                                                                              |
| Step 4: Show members of karmada                                                                                                                                              |
|                                                                                                                                                                              |
| (In karmada)~# kubectl  --kubeconfig /var/lib/karmada/karmada-apiserver.config get clusters                                                                                  |
|                                                                                                                                                                              |
├── —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— —— ┤
| Pull mode                                                                                                                                                                    |
|                                                                                                                                                                              |
| Step 1:  Send karmada kubeconfig and karmada-agent.yaml to member kubernetes                                                                                                 |
|                                                                                                                                                                              |
| (In karmada)~# scp /var/lib/karmada/karmada-apiserver.config /var/lib/karmada/karmada-agent.yaml {member kubernetes}:~                                                       |
|                                                                                                                                                                              |
| Step 2:  Create karmada kubeconfig secret                                                                                                                                    |
|  Notice:                                                                                                                                                                     |
|    Cross-network, need to change the config server address.                                                                                                                  |
|                                                                                                                                                                              |
| (In member kubernetes)~#  kubectl create ns karmada-system                                                                                                                   |
| (In member kubernetes)~#  kubectl create secret generic karmada-kubeconfig --from-file=karmada-kubeconfig=/root/karmada-apiserver.config  -n karmada-system                  |
|                                                                                                                                                                              |
| Step 3: Create karmada agent                                                                                                                                                 |
|                                                                                                                                                                              |
| (In member kubernetes)~#  MEMBER_CLUSTER_NAME="demo"                                                                                                                         |
| (In member kubernetes)~#  sed -i "s/{member_cluster_name}/${MEMBER_CLUSTER_NAME}/g" karmada-agent.yaml                                                                       |
| (In member kubernetes)~#  kubectl create -f karmada-agent.yaml                                                                                                               |
|                                                                                                                                                                              |
| Step 4: Show members of karmada                                                                                                                                              |
|                                                                                                                                                                              |
| (In karmada)~# kubectl  --kubeconfig /var/lib/karmada/karmada-apiserver.config get clusters                                                                                  |
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

File: pkg/karmadactl/karmadactl.go [hunk 2]
```diff
@@ -3,6 +3,7 @@ package karmadactl
 import (
 	"flag"
 	"fmt"
+	"github.com/karmada-io/karmada/pkg/karmadactl/cmdinit"
```
Comment (RainbowMango) [inline pkg/karmadactl/karmadactl.go:6]:
```
pkg/karmadactl/karmadactl.go:6: File is not `goimports`-ed with -local github.com/karmada-io/karmada (goimports)
	"github.com/karmada-io/karmada/pkg/karmadactl/cmdinit"

```
We usually group imports by `std packages` --> `third-part packages` --> `karmada packages`.

It's not big deal, but CI checks that...

Comment (RainbowMango) [reply pkg/karmadactl/karmadactl.go:6]: not fixed.

Comment (prodanlabs) [reply pkg/karmadactl/karmadactl.go]: Sorry, it has been fixed

**Comment by RainbowMango** [comment]
/assign
I'll try again against this latest patch.

**Comment by RainbowMango** [comment]
@lonelyCZ 
I'm going to merge this, let's iterate it by following PRs.

**Comment by lonelyCZ** [comment]
> I'm going to merge this, let's iterate it by following PRs.

Great, I also think so. Thanks.

**Pull request merged**